### ISSUE TITLE
feature request: keeping system dns unchanged

### ISSUE THREAD

**Issue by codl** [mullvad/mullvadvpn-app#473]
hi, I have a local DNS resolver for caching reasons and I want to keep using it when I enable mullvad, but the daemon changes `resolv.conf` without asking and even monitors it and reverts it every time it is changed. I couldn't find a way to disable this

could an option be added to disable this behaviour?

**Comment by faern** [comment]
We have plans on adding a feature to make the app set a custom DNS resolver. But I'm unsure we want to disable its DNS setting/monitoring completely. The app is fully responsible for the system security and our job is to try to protect the privacy of the people using it. Leaking DNS is terrible for privacy, so it's not something we want to make it too easy for users to do.

**Comment by codl** [comment]
well, a custom DNS would fix my problem either way

**Comment by akej74** [comment]
Hi, I'm using a Raspberry Pi with [Pi-Hole](https://pi-hole.net/) as a DNS server to filter out ad-related lookups. I have my router configured to set the IP of the RPi as DNS on my network over DHCP, but when starting the Mullvad client, the DNS is changed to e.g. 10.11.0.1.

An option to configure the DNS manually would be very much appreciated!

Side note/question: It using the Mullvad DNS at 193.138.219.228 as safe as using the one available on each VPN server (e.g. 10.11.0.1)?

**Comment by faern** [comment]
@akej74 Yes, the public DNS on that IP is still supported. As explained in our DNS leak guide (under Other users) all requests to it will actually be hijacked and redirected to the DNS resolver running on each VPN server when you are connected to Mullvad. So in practice, using `193.138.219.228` as your DNS through the tunnel effectively becomes equivalent to using the DNS available at the VPN server you are connected to.
https://mullvad.net/en/guides/dns-leaks/

**Comment by faern** [comment]
It is on the roadmap. But I don't have a time frame for it currently. It's not part of what we are working on at the moment.

**Comment by c-edw** [comment]
I've just spent 5 minutes or so trying to figure out why I couldn't reach any websites. Of course, I eventually checked the `resolv.conf` and realised that the Mullvad client must have changed it.

I'm alright with Mullvad changing it, but there needs to be some kind of notice to users that if the VPN is unexpectedly killed, the `resolv.conf` won't be changed back. In fact, why can't the original DNS configuration be stored somewhere, and have `mullvad-daemon` change it back when the VPN is not active?

**Comment by faern** [comment]
`mullvad-daemon` does restore the DNS when it's instructed to disconnect. What is it that you mean is unexpectedly killed? `mullvad-daemon` or OpenVPN? If OpenVPN dies unexpectedly then `mullvad-daemon` will directly try to start it again. Killing OpenVPN will not make the Mullvad VPN become disconnected, it will just make it retry connecting. For `mullvad-daemon` to stop trying and restore your system settings to use the internet unencrypted, you need to instruct Mullvad VPN to disconnect.

If you mean that `mullvad-daemon` is unexpectedly killed and that does not restore your settings then I say it depends a lot on the circumstances and your platform etc.

**Comment by c-edw** [comment]
My laptop battery died while I was connected. I guess it could be considered a bug - it depends on whether the client is supposed to restore the DNS settings in this case.

**Comment by faern** [comment]
@c-edw It should indeed have restored the DNS automatically on reboot. We have four different ways of managing DNS depending on what services are available on your distro etc. If you can reproduce the problem it would be awesome if you could send a problem report to our support through the app.

**Comment by Rouzax** [comment]
+1 for me.
Because it overrides my local DNS server I loose all connectivity to the other servers in my local domain which results in me being unable to logon (AD Domain), unable to access file shares, etc.
I understand why you would want to to do this but in my use case it makes the service useless.

**Comment by aalhitennf** [comment]
+1

Really needed feature for pi-hole users.

**Comment by Rouzax** [comment]
You can work around this by installing OpenVPN and creating connection files through Mullvad.

**Comment by iosadosd-quarto** [comment]
+1
I'm a Pi-Hole user and I'd love to be able to use it together with the (awesome) Mullvad app.
Until then, I guess I'll work around this by using OpenVPN.

**Comment by jelbo** [comment]
So how do I set up my OpenVPN clients to use a custom, local DNS server (Pi-hole)?

**Comment by fooness** [comment]
> This is how I set up my OpenVPN client configuration to use a custom, local DNS server (Pi-hole):
> 
> ```
> #block-outside-dns
> pull-filter ignore "dhcp-option DNS"
> dhcp-option DNS <local dns ip>
> ```

Anyone knows if the same works for OpenVPN Connect on iOS? Where `local dns ip` means local network and not localhost. (I want to use my PiHole DNS server when at home.)

Download the `.ovpn` files from [Mullvad’s config page](https://mullvad.net/en/download/config/), open in editor, add these two lines, securely send the files to iOS device … tried it, cannot really confirm if it works … but seems it doesn’t.

**Comment by DjCrays** [comment]
+1, would love if this is also possible in Wireguard and not only OpenVPN

**Comment by juliangaal** [comment]
+1, this feature would be killer for PiHole users!

**Comment by sheevy** [comment]
+2 if Mullvad could provide ads blocking DNS on their side insead of relying on Pi-Hole

**Comment by semente** [comment]
Would be great if it supported DNS-over-TLS, but I guess it must be implemented on Wireguard first.

**Comment by semente** [comment]
I've just figure out how to use DNS-over-TLS on a Wireguard connection:

(It may not apply to Mullvad apps, only official Wireguard software for GNU/Linux and Android)

On GNU/Linux system you must install and setup unbound or stubby software with the DNS-over-TLS service of your choice (e.g NextDNS.io), then set your /etc/resolv.conf to 127.0.0.1 and remove the DNS option from your Wireguard configuration (or just set it to `DNS = 127.0.0.1`)

On **Android 9 or later**, use the Wireguard official app to connect to Mullvad. Set the the DNS option of the desired VPN configuration to _blank_. It will make it use Android system's DNS. Go to Settings > Network > Advanced > Private DNS and set it to the DNS-over-TLS service of your choice.

I would suggest Mullvad developers to provide an option to "Use system DNS" on their apps. Thanks

**Comment by tobias-kuendig** [comment]
Having the client prepend a `nameserver 127.0.0.1` to `/etc/resolv.conf` would already help to prevent killing many automated development environments that run a local DNS resolver and would also provide a "hook" for more advanced user to use for their custom DNS needs.

Correct me if I'm worong but adding localhost as a DNS sever usually should not leak any DNS to the outside world.

**Comment by FilipoMoake** [comment]
Option to personalize DNS resolver is definitely something missing in Mullvad VPN, particularly if you you want to use DNS including ADBlocking.
I actually use Mullvad for the Wireguard compatibility and the hope of the DNS resolver option, but if this option is not implemented soon, sure that i will swap to another VPN provider, sorry to say that.
Is it something so hard to implement ?
Thanks in advance for understanding,
In the hope you'll solve that soon ...

**Comment by techwoes** [comment]
+1 for this option in order to allow pi-hole or other custom ad blocker. thanks for the hard work mullvad team

**Comment by juliangaal** [comment]
If you're using wireguard, shouldn't it be as easy as setting `DNS` in the client config to a custom server? They are normally in `/etc/wireguard/*.conf`, at least if you setup wireguard yourself. Not sure about where the mullvad app places the configs

**Comment by techwoes** [comment]
+1 on this for sure

**Comment by p1r473** [comment]
+1 here, cant use Mulvad while using Pihole

**Comment by MitchellCash** [comment]
This would be a great feature. I am keen to use NextDNS alongside the Mullvad VPN app. I could use the Wireguard app to set a different DNS resolver, but the functionality inside the Mullvad app would be my preference.

**Comment by funkythings** [comment]
> +1 here, cant use Mulvad while using Pihole

Are you using wireguard with mullvad? I think there may be a way to get to use both

**Comment by scafroglia93** [comment]
+1

**Comment by apptester235** [comment]
Would be great to see a Custom DNS setting in the app overall, Android, Desktop, etc. After such a long time it would please a lot of people to be able to use what ever custom DNS they want, self hosted or not. With features which the Mullvad DNS does not frankly provide, e.g. ad blocking, or domain white listing or black listening which other solutions like Pi-Hole can offer.

**Comment by p1r473** [comment]
> > +1 here, cant use Mulvad while using Pihole
> 
> Are you using wireguard with mullvad @p1r473 ? I think there may be a way to get to use both

I cant connect to Mullvad while Pihole enabled.
I got an error about DNS System Preference couldn't be changed.

**Comment by p1r473** [comment]
![image](https://user-images.githubusercontent.com/9235633/79001330-454a0100-7b1c-11ea-87da-7331a37acf88.png)

**Comment by benpro** [comment]
Interesting behavior:

On my One Plus 6 smartphone, I can use Mullvad app and private DNS of Android.
On my Samsung Galaxy Tab S5e I cannot use private DNS of Android if I use Mullvad app.

Why these differences... Intriguing.

**Comment by tazjin** [comment]
> But I'm unsure we want to disable its DNS setting/monitoring completely

This should be an option. For example, DNS settings are not mutable on my machine and I need to either patch the mullvad client or present it with a "fake" resolved to make it work. In practice that means I just don't use the client.

**Comment by ph00lt0** [comment]
@fooness using pihole as DNS is of course a good idea, but think about this. In an ideal senario when you are outside Mullvad will fall back to it's own DNS, however what happens if the network you are on uses the same local IP for it's DNS services as your pihole at home, then without you noticing you might leak your DNS data.

**Comment by husim0** [comment]
Would be great to have this option to set a custom DNS server. Thank you !

**Comment by jowabels** [comment]
+1, Pi-hole and Mullvad (Wireguard) would be lovely. If Mullvad DNS servers integrate with Pi-hole (I don't know if that is possible), that would be better.

**Comment by ph00lt0** [comment]
Actually what @jowabels you say would be great. I suggested to add a firewall option. They think that it's out of scope but honestly I should not be to difficult as I see it. Adding Pi Hole doing so would resolve the issue. I think that people would be willing to pay for that service, as least I would. Private Internet Access also has some features in blocking some ads (not that great but it's a start).

**Comment by RandomUserName22** [comment]
Custom DNS has many uses in all environments.  We have been waiting for it for a while now.  SHould be easy enough to implement - it can be done with manual port settings (you know the ones), so why not give those ports as a choice within the apps??????

**Comment by faern** [comment]
It's not just about using different ports on the VPN servers. We have very strict firewall rules and other blocks in place locally on the device running the app. These limit exactly which DNS server is and can be used. In order to implement custom DNS we need to add an input field in the app where the user can specify which DNS server they want and then all security rules must adapt to conditionally use this custom IP instead of the one for our server.

So far not very hard to implement. But the tricky part come as soon as we need to support custom DNS on either the LAN or on the routable internet, that's more tricky changes that I won't go into the details of.

All of this is simple in "vanilla" VPN apps that just create a tunnel, because they don't take measures to lock down unwanted stuff. We do, and as such all of those components must be able to adapt to changes in DNS, we can't just let any DNS out.

**Comment by RandomUserName22** [comment]
I might have oversimplified a little, but conceptually it is quite doable: you have competitors' apps which already do the same (expressvpn, nordvpn), and it can also be done with third-party apps (Viscosity can be set up to use custom dns, split tunnels, and with a little scripting can lock down the system so that if the tunnel drops all internet activity is blocked).

This is a feature that many people desire.  I know I would be willing to pay extra for it.

**Comment by jowabels** [comment]
Hi @faern quick question, does the DNS servers of Mullvad implement some sort of ad blocking? If no, will there be efforts to go into that direction?

**Comment by faern** [comment]
> Hi @faern quick question, does the DNS servers of Mullvad implement some sort of ad blocking? If no, will there be efforts to go into that direction?

They do not block anything. Yes there are internal discussions about supporting this sometime in the future. But that's an infrastructure question and not related to custom DNS in the app, even if the use cases can overlap for some scenarios.

**Comment by jamesmacwhite** [comment]
I've been recently evaluating Mullvad and found it to be excellent, except this missing feature in the app. It would be nice if the app allowed you to use local DNS, rather than forcing Mullvad DNS servers without any option to disable this. I understand your primary goal is privacy, but this is something more technical users would be aware of. Chances are their local DNS/custom DNS won't be a prying ISP and is probably something like PiHole with DNSCrypt or DNS over TLS, so still strong, even if it isn't going through the VPN tunnel.

I guess I could run OpenVPN directly with a Mullvad config to have local DNS still, but the app does offer a few good features which more easily configurable as a client, than having to go the OpenVPN config route. Also from a Linux perspective NetworkManager and it's OpenVPN plugin/functionality is really hit and miss, so I'd rather not have to do this.

Just my thoughts.

On an unrelated note, well done for being a VPN provider who actually offers IPv6, rather than the IPv6 leak protection crap others do. Thank you for being in the 21st century.

**Comment by mike386** [comment]
The ability to use a filtering DNS server is not just a nice addition to a privacy-oriented service, it is a must.

While preventing DNS leaks, the service also does not let the user stick with third-party services like AdGuard DNS - it just intercepts all DNS queries and redirects them to its own, non-filtering DNS server (also true if the user connects with WireGuard configuration file). To work this around the user may configure a local resolver that forwards all DNS queries to a filtering DNS server via DoT/DoH (and most probably configure that server to support such protocols). On every device.

Having to deal with ad-blocking browser extensions has its drawbacks. First, it is another piece of software with its vulnerabilities, often not receiving as much attention as browsers themselves (search for "Mega.nz Chrome Extension Hacked"). Second, it does not protect privacy device-wide (which is a big deal for mobile devices).

I hope that Mullvad will eventually have its filtering DNS server (10.9.0.2 would be a great option!) that would implement similar functionality to Pi-Hole.

**Comment by harish2309** [comment]
if you still want to use your own dns, like a Pihole IP, then I found a hacky solution to have mullvad stop editing resolv.conf. 

`chattr +i /etc/resolv.conf
`

So my DNS remains in place and Mullvad starts, however yes DNS will be leaking via this methods, which may or may not outweigh using your own DNS while connected to Mullvad.

**Comment by jamesmacwhite** [comment]
Yes. I guess you could essentially lock the file from being edited as a workaround, but ideally the app itself should allow you to maintain local DNS resolution. Someone else in this thread mentioned they want it for local DNS resolution i.e. Active Directory and therefore need local DNS, I am in a similar boat.

**Comment by aaronraimist** [comment]
@harish2309 at least for me on macOS (using `sudo chflags schg /var/run/resolv.conf`) that doesn't work. Mullvad just blocks DNS requests so nothing loads.

**Comment by jamesmacwhite** [comment]
@aaronraimist @harish2309 Same thing on Linux as well. The app has been designed for specifically force DNS requests, blocking the modification of resolv.conf from being modified just breaks DNS resolution entirely for me. It is not a solution.

The only viable way is running Wireguard or OpenVPN directly without the app.

**Comment by Mikaela** [comment]
This started affecting me yesterday as my Debian Testing's glibc updated to 2.31-1 which disables passing AD flag to software unless `/etc/resolv.conf` includes `options trust-ad` which generally should only happen when there is a local DNSSEC validating resolver running so now I don't get automatic validation of SSHFP records.

I would like the ability to trust systemd-resolved and a DoT server instead of Mullvad's in general also. I think the best case would be if Mullvad had public DoT server accessible also without Mullvad being enabled, so I could use the same configuration regardless of whether I had Mullvad enabled at that moment.

**Comment by luke-simo** [comment]
I love Mullvad. The fact that I can get a VPN service without it storing any identifying information is amazing.

However, blocking ads and tracking urls is more important for security and privacy than using a VPN. So without this feature I'm going to have to stop using Mullvad and move to another service which supports both :(

**Comment by erehwonstudios** [comment]
+1 : much love to all of you at mullvad, i've been a supporter from the beginning, but allowing us (if not you) to implement DNS filtering is a must. i have devices across every platform and, yes, i have the technical skills to jury rig a solution for some of them but, frankly, i just don't have the time. please, consider a priv DNS option (covered with security warnings and disclaimers), i'm about to drop money on a competitor that i'd much rather give to you.

**Comment by atmosblack** [comment]
I couldn't have said it better than the last 2 comments. Mullvad is my all-time favorite VPN, but I'd like to use Adguard Home as DNS server on my VPS to block ads etc on all my devices. Setting my own custom DNS has absolute priority for me. Since this thread is going on for years, I'm for sure not the only user wanting this feature.

**Comment by faern** [comment]
We know this is a highly requested feature. This is currently blocked on having the infrastructure support *not* hijacking all sent DNS. After that is done, we can allow the app to set a custom DNS.

**Comment by Le-Bowski** [comment]
@faern I for one am happy that you guys are being careful with this feature - although, don't get me wrong, I am very keen on seeing this option added to Mullvad.

**Comment by jamesmacwhite** [comment]
I guess because it is a common feature seen across a lot of VPN providers that is why there is a fair bit of noise around it, but can't argue with careful and calculated implementation, given DNS leaks are a big problem and very easy to do, which I guess it the main concern here to protect the innocent!

**Comment by RandomUserName22** [comment]
> We know this is a highly requested feature. This is currently blocked on having the infrastructure support _not_ hijacking all sent DNS. After that is done, we can allow the app to set a custom DNS.

Why cannot the app be allowed to use ports 1400/1401?

**Comment by faern** [comment]
That would be a very messy solution. It would make the custom DNS feature entangled in unrelated things, such as the relay selection parameter algorithm. And it would not work for users who need to use specific ports to escape their network anyway. So instead we wait until the infrastructure team allows enabling and disabling DNS hijacking per connection. It will take a bit longer, but the end result will be so much smoother and cleaner for users.

**Comment by itsmekevin212124** [comment]
I want to pile on a bit here as well.

Resetting the resolv.conf file basically breaks DNS for anyone who needs anything custom with DNS (kinda what it does), so anything with Pi-Hole, OpenNic, NameCoin, businesses that require custom DNS, ect. simply don't work with the current Mullvad app. Therefore, although it may seem like a good security feature this app also breaks workflow for many users.

Just wanted to point that out for anybody who mat be wondering why the ability to use custom DNS is important. :P

________________________________________________

**Edit**: I did also want to say, this might be a good feature to put under the "advanced" section.

**Comment by dhaavi** [comment]
I just investigated why the Mullvad App could not connect when the Portmaster Application Firewall is running. It turns out that the Mullvad App requires the Windows DNS Client (the `dnscache` service) to be running, which the Portmaster disables to gain its required system integration.

I would be great to not have the Mullvad App depend on this service, so that all Portmaster users can also use Mullvad.

I can confirm that Mullvad works perfectly if I use the OpenVPN profiles. As Portmaster uses encrypted DNS, there is no DNS interception possibly on the VPN server - I understand that this is something that is going on from one of the comments. Although I'd like to understand why this is done.

**Comment by faern** [comment]
We very carefully filter DNS requests in the client app and hijack them on the servers for privacy reasons. DNS is giving away basically all details about what you do online. So having a VPN tunnel without properly handled DNS is kind of as pointless as not having a VPN tunnel at all, from a privacy perspective.

Maybe the users of certain other third party software are safe without us handling DNS. But in a default installation they are not, so we must take matters into our own hands to guarantee privacy.

All of this was designed before encrypted DNS was a thing. The encrypted DNS "revolution" really cause problems for privacy concious VPN users. They don't need it. The best they can do (in a Mullvad setup at least) is to ask the VPN server inside the tunnel anyway. Modulo ad blocking or other special handling of requests of course. I'm not talking about that here. The app makes sure that all DNS requests go encrypted in the tunnel.

The server side hijacking specifically is designed for users not using our app, but using vanilla OpenVPN or WireGuard. Since their devices might send DNS requests far and wide, we intercept them and make sure they don't leave the server. That way their privacy is kept a bit more intact. But we are working on making this optional. So people can disable this hijacking for their connections. When this is available the app will do it automatically and handle DNS privacy on the client side, allowing a custom DNS to be used if desired.

**Comment by dhaavi** [comment]
Thanks for the detailed explanation. It is true that DNS is quite tricky when it comes to privacy. Encrypted DNS without a VPN provides a decent amount of security, but not all that much privacy.

Intercepting stray DNS queries is actually a very nice protection against misconfiguration. 👌  Haven't seen this in a VPN before. This adds great value to non-technical users and is vital in bringing more privacy to everyone. In fact, the Portmaster does the same, but locally.

Looking forward to more compatibility when this feature launches!

**Comment by atmosblack** [comment]
I guess what some consider a bug, for others can be a feature. ;) I was fighting with DNS/VPN settings on iOS over the past days. While on Android it simply works to have an encrypted DNS parallel to a VPN, on iOS the VPN always overwrites and wins over the encrypted native DNS. The only possibility was to connect via IKEv2, have Adguard Pro in split tunnel mode and use my NextDNS settings there to be able to run both DNS & VPN at the same time. It doesn't work with Wireguard (or OpenVPN) unfortunately. 
Very few VPN clients have a custom DNS option (IVPN, PIA), but then you still have to manually register your IP in nextDNS everytime, which is annoying. That's just iOS though...

What makes it even more annoying is that Mullvad hasn't even managed to give their users a custom DNS option that works on all platforms! So you don't get to the even more advanced challenges with Mullvad.

Until they manage that, I'll stay with PIA - knowing very well, that it's located in the US and has not the best leadership out there, but from a technical perspective it's still the best option if you are an advanced user.

**Comment by faern** [comment]
> Mullvad App requires the Windows DNS Client (the dnscache service) to be running, which the Portmaster disables to gain its required system integration.

This is interesting. We don't directly depend on `dnscache`. But I assume you are referring to our use of `netsh` to set DNS servers on the virtual interface then? I guess that command requires `dnscache` to be running to actually work.

**Comment by kayg04** [comment]
> but then you still have to manually register your IP in nextDNS

I wish they supported ipv6 addresses. As you said, with ipv4, you'd have to register with another service for dynamic DNS. All of that is very inconvenient.

> What makes it even more annoying is that Mullvad hasn't even managed to give their users a custom DNS option

If you're on Windows, I currently have a hack running that works. Use YogaDNS to configure NextDNS and have it running alongside the Mullvad app. This leads to your system using NextDNS + Mullvad's DNS (below is a result of both of them connected simultaneously). All the blocked stuff seems to respond correctly.

 ![image](https://user-images.githubusercontent.com/30401343/94584890-e41a1380-029c-11eb-9669-eba15e761342.png)

![firefox_Eh0TtLjPLb](https://user-images.githubusercontent.com/30401343/94585097-280d1880-029d-11eb-82bd-5c3615a10262.png)

**Comment by dhaavi** [comment]
> ... use of `netsh` to set DNS servers on the virtual interface then? I guess that command requires `dnscache` to be running to actually work.

Yes, that is also how I understood it. `netsh` reported that the service was not running.

**Comment by RandomUserName22** [comment]
Any ETA on custom dns?  I may have to abandon Mullvad until this feature is available.

**Comment by arik-gamerlink** [comment]
Am I incorrect to assume that this [PR](https://github.com/mullvad/mullvadvpn-app/pull/2096) addresses this? At least on Android.

**Comment by kayg04** [comment]
I realize that isn't exactly a fix but if you use the Private DNS feature in settings on Android along with the Mullvad app to connect to the VPN, everything works as intended.

07-Oct-2020 04:15:12 arik-gamerlink <notifications@github.com>:

**Comment by matthowell43** [comment]
Just signed up to Mullvad and was disappointed to see this feature isn't available. I feel it should be in the 'Advanced' section of the app, perhaps with a warning about the security risks.

**Comment by HK53** [comment]
Yesterday I installed NextDNS. It seemed to work alright,  but after sleepmode of my laptop Mullvad did not connect because it could not set de DNS. So I am in the ranks.

I found a work around. I stopped NextDNS in the terminal, then reconnected Mullvad with success, and restarted NextDNS. A little bit of a hassle, but it works. No problem to do this for a short while.

Today I installed the recent 2020.8-Beta2 and there is an option to add a custom VPN, but it needs a local ip address.

Am I correct that this custom VPN is (or will become) a solution to the conflict with NextDNS? If so, is there a local ip address to direct DNS queries to NextDNS? And will the NextDNS deamon use the Mullvad tunnel?

**Comment by cwmke** [comment]
> Yesterday I installed NextDNS. It seemed to work alright, but after sleepmode of my laptop Mullvad did not connect because it could not set de DNS. So I am in the ranks.
> 
> I found a work around. I stopped NextDNS in the terminal, then reconnected Mullvad with success, and restarted NextDNS. A little bit of a hassle, but it works. No problem to do this for a short while.
> 
> Today I installed the recent 2020.8-Beta2 and there is an option to add a custom VPN, but it needs a local ip address.
> 
> Am I correct that this custom VPN is (or will become) a solution to the conflict with NextDNS? If so, is there a local ip address to direct DNS queries to NextDNS? And will the NextDNS deamon use the Mullvad tunnel?

For NextDNS I believe you would use 127.0.0.1 as the custom option.

**Comment by HK53** [comment]
thank you, @cwmke !

I tried 127.0.0.1 (a wild guess), but Mullvad immediatly blocked, saying 'cannot set dns'.

My problem is also that I have no idea how all these network component relate to each other (native Linux/Ubuntu, with KDE Gui on top, Mullvad, NextDNS)...

regards, H

**Comment by pinkisemils** [comment]
I installed the NextDNS client on debian-testing with our client 2020.8-beta2 and tested it out via their test [server](https://test.nextdns.io). It works just fine as long as you add `127.0.0.1` to the custom DNS servers.
![image](https://user-images.githubusercontent.com/7757572/102381406-121c1380-3fc1-11eb-8a26-ca1a4bf53899.png)
 It might help to disable `systemd-resolved` as it's listening on port 53, the same port NextDNS's client wants to use - 
```bash
sudo systemctl stop systemd-resolved
``` 

If you're using NextDNS to use DoH or DoT, you could also just use [`systemd-resolved`]  - go to the setup page of your [nextdns](https://my.nextdns.io) account to get the config file for you. Then, enable, run the following commands as root:
```bash
systemctl enable systemd-resolved
systemctl start systemd-resolved
rm /etc/resolv.conf && ln -s /var/run/systemd/resolved/resolv.conf /etc/stub-resolv.conf
```

**Comment by HK53** [comment]
Thank you @pinkisemils !

I tried "sudo systemctl stop systemd-resolved". That does not help. After entering 127.0.0.1 as custom DNS Mullvad blocks. After stopping NextDNS Mullvad agrees to reconnect. After Start NextDNS it stays connected.

Your followup advice is very condensed... I needed to read it several times.

I installed NextDNS already with sh -c "$(curl -sL https://nextdns.io/install)", that was their recommended install.

test.nextdns.io says 'status ok' and 'protocol DOH'.

I copied from my.nextdsn.io/setup the lines under "systemd-resolved" into /etc/systemd/resolved.conf.

I do not know what you mean by 'enable'. I assumed stop and start ... Mullvad? NextDNS? I did both.

I exucuted the three commands you mentioned with sudo. the 'ln -s ..." needed an extra sudo...

Now, after all this done, I still have the problem that Mullvad blocks after resume from sleep. And still the work around of  temporarily stopping NextDNS helps.

Please note, my problem is not that NextDNS is not working. My problem is that Mullvad blocks after sleep.

Now what shall I do, or what needs to be done? For me the situation is equal to not using a custom DNS. Should I perform any command to return something to previous setup?

Thanks for your help.

**Comment by pinkisemils** [comment]
With the `systemd-resovled` use, I meant to only use `systemd-resolved` without the NextDNS client. You'd still be using their service, just without their client. So, firstly you'd have to remove or disable the NextDNS client, and then re-enable systemd-resolved. 

Regardless, would you mind posting the contents of your `/etc/resolv.conf` when the daemon fails to apply DNS? Feel free to censor any addresses, just want to see the arguments used there.

**Comment by HK53** [comment]
thank you again, @pinkisemils

somehow I managed to get Mullvad and NextDNS living peacefully together, even after resuming from system sleep... 

I tried to enable systemd-resolved with NextDNS uninstalled, but that resulted in not being able to load any new webpages at all (no DNS?)

I re-installed NextDNS => new webpages again (relief)

Then, onetime, at resume from system sleep, Mullvad kicked out NextDNS... I noticed this because of a status change on my.nextdns.io. The roles were changed?

Anyhow, at this moment, both Mullvad and NextDNS just continue after resume from system sleep. So my initial problem for posting seems to have been solved.

I would like to understand better what has happened...

Question: do I need to clear confic slack w.r.t. systemd-resolved ?

Question: tcpdump does not show any DNS traffic on port 53. Might that mean only now systemd-resolved is effectively stopped? (trying to figure out how this all relates)

Question: are you still interested in some information?

**Comment by pinkisemils** [comment]
One reason why DNS didn't work via systemd-resolved might be because you didn't link it's resolv.conf to `/etc/resolv.conf` via `ln -s /var/run/systemd/resolv/resolv.conf /etc/resolv.conf`. But you don't have to worry about that.

> Question: do I need to clear config slack w.r.t. systemd-resolved ?
You can remove the config changes you made if you're not going to use it just for clarity.
> Question: tcpdump does not show any DNS traffic on port 53. Might that mean only now systemd-resolved is effectively stopped? (trying to figure out how this all relates)
If you've successfully enabled NextDNS client, it will probably use DoH or DoT, which use different ports.
> Question: are you still interested in some information?
I'd be very interested in your `/etc/resolv.conf` if our app still fails to set DNS.

**Comment by HK53** [comment]
@pinkisemils everyting is ok now. Many thanks for you assistance!

**Comment by HoboDev** [comment]
I'm trying the new beta for Linux. Unfortunately I can not start the gui because my Krunner is broken and I don't know how to start it from command line.
But when I set a custom dns via cli It overrides the mullvad dns. Is there any options to use it side by side? (Currently I use a script which sets custom-dns to off gets the dns from resolv.conf and sets new custom-dns with both the custom dns and the mullvad dns)

Two things come to my mind: It would be nice to have an option for getting the nameserver without script hackery. Also a hook for a script that runs after connecting.

**Comment by faern** [comment]
> I'm trying the new beta for Linux. Unfortunately I can not start the gui because my Krunner is broken and I don't know how to start it from command line.

The GUI start script is at `/opt/Mullvad\ VPN/mullvad-vpn`, so you can just run that.

@HoboDev What's your use case for this? That's not what custom DNS was designed for. Custom DNS is for when you don't want to use our DNS resolver in the tunnel. Because you have a special DNS that does ad blocking or where you have some special local subdomains or something. So yes, it switches between the Mullvad one or your custom one on purpose.

**Comment by Mattrazol** [comment]
Is this feature going to be released for android as well?

**Comment by faern** [comment]
Yes. Custom DNS will be on Android and desktop.

**Comment by Mattrazol** [comment]
@faern When will that be available?
And if for example I have a Pi-Hole and I set it as my primary DNS will I be able to leave the local IP address of my Pi-Hole as primary and have the secondary handled by the app itself so when I'm not connected to my WiFi network it will fallback on the Mullvad one? I hope that this will makes sense..

**Comment by faern** [comment]
I don't have an ETA for the next app release. It depends on how well development and internal testing goes. Probably within a few weeks.

No it's not really designed to have a fallback. You either configure it to have a custom DNS **or** you let it manage DNS internally automatically. Not both. The app can't determine if your pi-hole is available or not in any good way.

**Comment by Mattrazol** [comment]
@faern what about querying the first and if it doesn't find the domain or it fails than fallback on the secondary (if set) and in case if is not set use the default one from Mullvad?

**Comment by p1r473** [comment]
> > > +1 here, cant use Mulvad while using Pihole
> > 
> > 
> > Are you using wireguard with mullvad @p1r473 ? I think there may be a way to get to use both
> 
> I cant connect to Mullvad while Pihole enabled.
> I got an error about DNS System Preference couldn't be changed.

Support said I am required to use the Windows DNS Caching service, which I don't want, else I need to use OpenVPN
Sad, because I wanted to use the Mullvad Wireguard app and not have to use DNS Caching service.

**Comment by vr** [comment]
what about the "leave my dns unchanged" option ? my /etc/resolv.conf has immutable flag as purpose and i don't want any application to touch it, having to remove that to set again 127.0.0.1 as custom dns is not enough for me.

**Comment by sylv-io** [comment]
Is there a way to use systemd-resolved stub listener on `127.0.0.53` ?
I would like to be able to resolve hosts on my private network, using a custom dns server.

**Comment by faern** [comment]
:tada: Yesterday we released app version `2021.1` for Windows, macOS and Linux. That version supports custom (local) DNS :tada: 

https://mullvad.net/en/blog/2021/2/11/long-awaited-feature-introduced-desktop-app-20211/

This means that you can now go to Settings -> Advanced and set up custom DNS resolver IPs. However, since our VPN servers still hijack all DNS requests going through them only **local** resolvers are possible. This means you can use `127.0.0.53` for a resolver running on your system or use something like `10.0.0.1` for your router or something.

The same functionality will come for Android in the next release. We have no ETA for when that is. But expect it in a few weeks maybe.

That means most of the functionality in this feature request has now been implemented. I hope it will work well for all of you!

**Comment by fooness** [comment]
Slightly off-topic, but also somewhat related: Might there be any problems when using Mullvad’s DNS server (IP via: https://mullvad.net/en/help/dns-leaks/) as the upstream DNS of my raspberry pi (with pihole) to prevent DNS leaks? Are there more/alternate Mullvad DNS server IPs in case of downtime or lots of traffic?

**Comment by faern** [comment]
If you just use our public DNS server as your DNS resolver and don't have a VPN tunnel or anything, then your ISP or anyone else between you and our server will see your requests. Normal DNS is not encrypted. And if you do have a VPN tunnel on the other hand it's a different deal.

We have also just now released our own DoH and DoT solutions (still in beta). You can try that out if you are interested: https://mullvad.net/en/help/dns-over-https-and-dns-over-tls/

For other inquiries about DNS and usage of our infrastructure please contact support@mullvad.net instead as this is not app related.

**Comment by TheOriginalCoda** [comment]
I have lost hours pissing about with DNS today - when my server is connected to Mullvad all DNS lookups fail. I cant run incoming connections to my public IP either when mullvad is running, all attempts at setting up port forwarding on my account page fail whether using wireguard or openvn - this all works when mullvad is disconnected! Not only that but often a 'mullvad status' request says 'connection blocked... too many clients' which is bananas, and searching for help on the web is fruitless - it finally brought me here to find out that I cant use my local DNS I have set up already, and the new custom-dns settings in the linux client do not work. I've cancelled my recurring payment and will look for something else. Its a shame, I've been happy for years with mullvad but havent tried anything 'outside of the box' until now.

**Comment by faern** [comment]
@TheOriginalCoda You can set a custom DNS is the app. I'm not sure what part of it / why it does not work for you. However, if the local resolver is on `localhost` you would need some extra firewall trickery to allow it to reach the internet, since our app would be blocking anything from going outside the tunnel. And any DNS inside the tunnel is currently captured by our relay servers. You can't for example reach `1.1.1.1` inside the tunnel, because the relay would hijack the request. This latter problem is something outside of the control of the app. But the restriction will be lifted very soon. Then custom DNS can go through the tunnel!

As for incoming connections. Yes, our app blocks that as by design. When enabled nothing outside the tunnel is allowed. However, you can manually add firewall rules with higher priority than the rules our app add, and then allow whatever it is you want to allow there. See this comment for some inspiration: https://github.com/mullvad/mullvadvpn-app/issues/2097#issuecomment-799485645

**Comment by p1r473** [comment]
Can you please make this not require dnscache Windows service?

**Comment by dhaavi** [comment]
I'm not part of Mullvad, but wanted to add my .2c:

At some point with Windows 10 the dnscache Windows service became a required service. This is very poorly documented.
A hint for this is that you cannot turn it off anymore via the service manager, but have to manually edit the registry.

The dependance on dnscache goes so far that even Powershell commands require it to be running and just silently fail if it's not.
We ourselves built a firewall that required it to be turned off, which created huge problems. For example, Docker for Windows won't run with the new backend anymore.

My .2c as a developer from another project: It just does not make sense anymore to turn off dnscache. I feel that at this point, turning it off can be seen as a modified operating system.

**Comment by p1r473** [comment]
I personally dislike dnscache as I run my own DSN server, and during a troubleshooting or whitelisting issue, I have to disable the dnscache so I am getting proper nslookups, unhindered by the cache

**Comment by dhaavi** [comment]
In that case I would recommend to leave dnscache enabled and use the amazing tool `dig`. [Here is an install guide](https://www.configserverfirewall.com/windows-10/dig-command-windows/). This bypasses the Windows dnscache and does lookups directly.

Just be careful when using it while having Mullvad enabled, as - afaik - they catch astray dns queries and redirect them to themselves.

**Comment by p1r473** [comment]
Thanks, good idea.!

**Comment by faern** [comment]
We have stopped hijacking astray DNS requests **if you use our app and have a WireGuard tunnel**. Any WireGuard key uploaded by the app in the last few weeks has DNS hijacking disabled. This is to allow our custom DNS feature to work as expected. But it also means you can query whichever DNS server you want in the tunnel via dig :)

**Comment by ph00lt0** [comment]
@faern any plans to allow localhost with custom port on android as custom DNS? I didn't want to open a new issue yet, because I feel this is related. It would be great to be able to use [Nebulo](https://git.frostnerd.com/PublicAndroidApps/smokescreen/-/blob/master/README.md) (in non-vpn mode) to be able to filter network activity with block lists and afterwards route it to DOH. Allowing blocklists within mullvad would also do, but not sure if that is on the table.

**Comment by faern** [comment]
@ph00lt0 I can add to our research to check if we can support custom ports on Android easily or not. But this is nothing we currently have in our backlog or a timeline for.

I'm not sure exactly what it is that you want to block. But just so you know: We support ad and tracker blocking via DNS on all our relays already. You can use it on Android by setting a custom DNS. Read this blog post for the details: https://mullvad.net/en/blog/2021/5/27/how-set-ad-blocking-our-app/

**Comment by ph00lt0** [comment]
@faern thanks you.

I am aware that you offer ad- and tracker blocking already, but I would like to be in control over what I block, mainly to block more things, that might break some functionalities which I do not need and known malware lists.
My idea is to forward the DNS first to Nebulo, add my own blocklists there and then forward the dns request to mullvad's DOH. A native option to select blocklists and blocked domains in mullvad would be awesome, but I thought, I keep it simple ;)

**Comment by faern** [comment]
Very similar issue: #3173

**Comment by n-stone** [comment]
> We have stopped hijacking astray DNS requests **if you use our app and have a WireGuard tunnel**. Any WireGuard key uploaded by the app in the last few weeks has DNS hijacking disabled. This is to allow our custom DNS feature to work as expected. But it also means you can query whichever DNS server you want in the tunnel via dig :)

```
$ nslookup google.de
Server:		10.64.0.1
Address:	10.64.0.1#53

Non-authoritative answer:
Name:	google.de
Address: 142.250.179.195

$ nslookup google.de 1.1.1.1
;; connection timed out; no servers could be reached
```

<img width="308" alt="image" src="https://user-images.githubusercontent.com/31704359/154745523-10683538-53f1-43e9-ba7b-f11152605838.png">


@faern am I doing something wrong?

**Comment by pinkisemils** [comment]
> @faern am I doing something wrong?
You'll have to enable `1.1.1.1` as a custom resolver in the app still. Otherwise the firewall will drop DNS traffic.

**Comment by MahouShoujoMivutilde** [comment]
Is there a way to use default system dns for local domains (like `.local` or `.test` etc) and mullvad (`10.64.0.1`) for others, on linux (latest arch, systemd-resolved)?

I found that I can sort of make it work with dnsmasq

<details>
  <summary>config</summary>
  
```
# /etc/dnsmasq.conf
server=/test/10.0.0.1

# private mullvad
server=10.64.0.1

listen-address=127.0.0.1

proxy-dnssec
no-hosts
no-resolv

no-negcache
cache-size=10000

min-cache-ttl=300
max-cache-ttl=86400

```
</details>


and setting custom dns server in the app to `127.0.0.1` and `10.0.0.1`. But that's defeats the purpose, since now both of them will be used for everything.

If I add `127.0.0.1` alone - it doesn't work since 

```
> drill @10.64.0.1 google.com
Error: error sending query: Error creating socket
```

So apparently default firewall rules block even mullvad's own dns.

I tried to set custom dns to `127.0.0.1` and add a custom firewall rule inspired by [Split tunneling with Linux (advanced)](https://mullvad.net/ru/help/split-tunneling-with-linux-advanced/#allow-dns).

<details>
  <summary>nft rule</summary>
```
table inet excludeTraffic {
  chain excludeDns {
    type filter hook output priority -10; policy accept;
    ip daddr 10.0.0.1 udp dport 53 ct mark set 0x00000f41 meta mark set 0x6d6f6c65;
    ip daddr 10.0.0.1 tcp dport 53 ct mark set 0x00000f41 meta mark set 0x6d6f6c65;

    ip daddr 10.64.0.1 udp dport 53 ct mark set 0x00000f41;
    ip daddr 10.64.0.1 tcp dport 53 ct mark set 0x00000f41;
  }
}

```
</details>


This allows `10.0.0.1` to work, but `10.64.0.1` still gives the error above.

I think this happens because `table inet mullvad`  has

<details>
  <summary>chain output</summary>


```
chain output {
    type filter hook output priority filter; policy drop;
    oif "lo" accept

##### In iptables this would just accept and exit immediately from the chain,
    # but apparently that's not how nft works, so idk how to to do it here.
    ct mark 0x00000f41 accept 
....
    oif != "wg-mullvad" udp dport 53 ip daddr 127.0.0.1 accept
    oif != "wg-mullvad" tcp dport 53 ip daddr 127.0.0.1 accept

#### HERE I think is the problem: reject all dport 53 first and
    udp dport 53 reject
    tcp dport 53 reject with tcp reset

#### only then accept wg-mullvad output
    oif "wg-mullvad" accept
...
}

```

</details>

Which is weird - shouldn't it just accept everything that goes out to `wg-mullvad`? I don't see a way how that can lead to dns leaks. (Since it will go inside a tunnel anyway).

Any ideas? 

I want to use the app because it's convenient and has `mullvad-exclude`, but the firewall makes it very difficult to do.

**Comment by pinkisemils** [comment]
You can setup exclusion rules for your local resolver address and apply said rules on boot. All you have to do match your DNS traffic and assign an exclusion mark.  There's a handy guide for this on our [website](https://mullvad.net/en/help/split-tunneling-with-linux-advanced/).  If you exclude your local DNS traffic, `systemd-resolve` will be able to reach your local resolver. You should be able to configure `systemd-resolve` to only use that resolver for your local domains too. But do keep in mind that if you do this, our app won't be able to ensure that DNS traffic won't ever leak - then you're relying on `systemd-resolve` to not leak.

**Comment by MahouShoujoMivutilde** [comment]
Yes, I've seen this guide, even linked it myself in the original comment.

Anyway, I've managed to solve this.

So, the problem:

1. When custom dns is enabled even default mullvad's internal dns `10.64.0.1` is blocked by the firewall because `table inet mullvad chain output` has `... dport 53 reject` before `oif "wg-mullvad" accept`.

So any dns traffic not going to the custom dns is blocked.

2. It is easy to make something bypass the tunnel altogether (like I did with `10.0.0.1` above) - just set mark and meta mark.

But it is surprisingly hard to bypass the dns filter for traffic that will go _inside_ the tunnel (like `10.64.0.1`).

Seeing `ct mark 0x00000f41 accept` at the start of `table inet mullvad chain output` I've expected that setting the said mark (without meta mark) would be enough, but it isn't.

Ct mark here is not enough, only inserting `accept` for `10.64.0.1` dns in the mullvad's output chain will actually work.

What I did:

1. Just disable custom dns and apply this nft rule:

```
table inet excludeTraffic {
  chain excludeDns {
    type filter hook output priority -10; policy accept;
    ip daddr 10.0.0.1 udp dport 53 ct mark set 0x00000f41 meta mark set 0x6d6f6c65;
    ip daddr 10.0.0.1 tcp dport 53 ct mark set 0x00000f41 meta mark set 0x6d6f6c65;
  }
}
```

That way `10.64.0.1` is allowed by default and you don't need to try to insert appropriate `accept` rule in to existing `table inet mullvad chain output`.

2. Dnsmasq config is the same as above, I haven't figured how force systemd-resolved to use `10.0.0.1` only for my `.test` and `10.64.0.1` for others.

3. In `/etc/systemd/resolved.conf`

```dosini
[Resolve]
DNS=127.0.0.1
DNSStubListener=no
```

To force it to use dnsmasq globally.

Kind of clunky, would prefer if all of the dns traffic going through the tunnel would just be `accept`ed, but this works.

**Comment by Zephilinox** [comment]
this might help resolve issues using mullavad with portmaster which also enforces a DNS

https://github.com/safing/portmaster/issues/313
https://github.com/safing/portmaster/issues/445

**Comment by mietzen** [comment]
> > We have stopped hijacking astray DNS requests **if you use our app and have a WireGuard tunnel**. Any WireGuard key uploaded by the app in the last few weeks has DNS hijacking disabled. This is to allow our custom DNS feature to work as expected. But it also means you can query whichever DNS server you want in the tunnel via dig :)
> 
> ```
> $ nslookup google.de
> Server:		10.64.0.1
> Address:	10.64.0.1#53
> 
> Non-authoritative answer:
> Name:	google.de
> Address: 142.250.179.195
> 
> $ nslookup google.de 1.1.1.1
> ;; connection timed out; no servers could be reached
> ```
> 
> <img alt="image" width="308" src="https://user-images.githubusercontent.com/31704359/154745523-10683538-53f1-43e9-ba7b-f11152605838.png">
> 
> @faern am I doing something wrong?

This solved my Problem: 
`curl -sSL https://api.mullvad.net/app/v1/wireguard-keys -H "Content-Type: application/json" -H "Authorization: Token YOURMULLVADACCOUNTNUMBER" -d '{"pubkey":"YOURPUBLICKEY"}'`

https://schnerring.net/blog/use-custom-dns-servers-with-mullvad-and-any-wireguard-client/

With the API call above I got a "non DNS hijacking IP" and could finally use unbound + Mullvad Wireguard on my router.

**Comment by schnerring** [comment]
> https://schnerring.net/blog/use-custom-dns-servers-with-mullvad-and-any-wireguard-client/

Author of that blog post here 👋

I'm currently seeing DNS timeouts when using Mullvad servers provided by `31173 `. So far I tested `ch{5,6,7,8,9}-wireguard`, `de-fra-wg-004`, `gb13-wireguard`, `nl-ams-wg-002`, and `se-sto-wg-001` all with the same results. The behavior is the same whether I use custom DNS servers or not. The Swiss `DataPacket`, `M247`, and `PrivateLayer` servers that I tested work as expected.
 
1. [I tweeted @mullvadvpn](https://twitter.com/schnerringo/status/1643405620924542976)
2. One of my readers [reported the same issue in the comments of that blog post](https://schnerring.net/blog/use-custom-dns-servers-with-mullvad-and-any-wireguard-client/#remark42__comment-41e1b4b7-6218-4bed-b838-6f33de5dfba0). He also says:  
> But recently on every 3rd or 4th try I’ll get a hijacking IP again / unbound can’t reach the dns root servers.
3. Testing these servers with the **official Mullvad VPN App + Custom DNS Servers** returns the same issues
4. Testing this on another residential connection to rule out any ISP issues returns the same issues

Regarding (2): has the API endpoint to request non-DNS-hijacked IPs changed away from `https://api.mullvad.net/app/v1/wireguard-keys`?

**Comment by dlon** [comment]
@schnerring We do not use that endpoint anymore. You should probably disable DNS hijacking when creating a *device*, not when rotating the key owned by that device.

Untested, but this should work:

```
access_token=$(curl -X 'POST' 'https://api.mullvad.net/auth/v1/token' \
    -H 'accept: application/json' -H 'content-type: application/json' \
    -d '{ "account_number": "SOME_ACCOUNT_NUMBER" }' | jq -r .access_token \
)

# create a new device, and set hijack_dns to false
curl -X POST https://api.mullvad.net/accounts/v1/devices \
    -H 'content-type: application/json' -H "Authorization: Bearer $access_token" \
    -d '{"pubkey":"PUBKEY HERE","hijack_dns":false}'

# rotate key
curl -X PUT https://api.mullvad.net/accounts/v1/devices/{DEVICE ID HERE}/pubkey \                                                                   
    -H 'content-type: application/json' -H "Authorization: Bearer $access_token" \
    -d '{"pubkey":"NEW PUBKEY HERE"}'

```

**Comment by dmorin20** [comment]
Can this be achieved to create a wireguard tunnel on a dd wrt router?

**Comment by p1r473** [comment]
Local DNS Resolution has broken for me in build 2025.5
https://github.com/mullvad/mullvadvpn-app/issues/8392
